src/Chainweb/TreeDB.hs

```diff
@@ -290,83 +267,14 @@ class (Typeable db, TreeDbEntry (DbEntry db)) => TreeDb db where
     -- cursor. Implementations should block on IO at most a constant amount of
     -- time.
     --
-    -- The default implementation is based on 'allEntries', which in most cases
-    -- doesn't give good performance.
-    --
     entries
         :: db
         -> Maybe (NextItem (DbKey db))
         -> Maybe Limit
         -> Maybe MinRank
         -> Maybe MaxRank
-        -> S.Stream (Of (DbEntry db)) IO (Natural, Eos)
-    entries db k l mir mar = allEntries db Nothing
-        & applyRank mir mar
-        & timeoutStream 1000 {- microseconds -}
-        & void
-        & seekLimitStream key k l
-    {-# INLINEABLE entries #-}
-
-    -- | The infinite stream of all entries in ascending order. The stream may
-    -- block indefinitely.
-    --
-    -- The default implementation is based on @entries@ and spins in a busy loop
-    -- while waiting for new entries.
-    --
-    allEntries
-        :: db
-        -> Maybe (NextItem (DbKey db))
-        -> S.Stream (Of (DbEntry db)) IO ()
-    allEntries db = go
-      where
-        go x = do
-            n :> (_, eos) <- entries db x Nothing Nothing Nothing
-                & S.copy
-                & S.last
-            case eos of
-                Eos True -> error "code invariant violation"
-                Eos False -> go (Exclusive . key <$> n)
-    {-# INLINEABLE allEntries #-}
-
-    -- ---------------------------------------------------------------------- --
-    -- Leaves
-
-    -- | A set of leaves in ascending order.
-    --
-    -- The semantics are as follows:
-    --
-    -- A set of nodes is returned such that each node in the tree, that has a
-    -- rank within the given range of ranks, is either a successor or a
-    -- predecessor of the nodes in the set.
-    --
-    -- If no minimum rank is given, a minimum rank of zero is assumed. If no
-    -- maximum rank is given, a maximum rank if infinity is assumed.
-    --
-    -- Note that, if the minimum rank is zero, returning the genesis block is
-    -- always a legal result, but implementations should try to return a large
-    -- set.
-    --
-    leafEntries
-        :: db
-        -> Maybe (NextItem (DbKey db))
-        -> Maybe Limit
-        -> Maybe MinRank
-        -> Maybe MaxRank
-        -> S.Stream (Of (DbEntry db)) IO (Natural, Eos)
-    leafEntries db k l mir mar = leafKeys db k l mir mar
-        & lookupStreamM db
-    {-# INLINEABLE leafEntries #-}
-
-    leafKeys
-        :: db
-        -> Maybe (NextItem (DbKey db))
-        -> Maybe Limit
-        -> Maybe MinRank
-        -> Maybe MaxRank
-        -> S.Stream (Of (DbKey db)) IO (Natural, Eos)
-    leafKeys db k l mir mar = leafEntries db k l mir mar
-        & S.map key
-    {-# INLINEABLE leafKeys #-}
+        -> (S.Stream (Of (DbEntry db)) IO (Natural, Eos) -> IO a)
```

Comment:
Comment this change please

Reply:
I tried to avoid this change of the TreeDB API, but finally had to make it. 

RocksDB iterators must be allocated and deallocated explicitly. The `Stream` type doesn't support resource management within the stream, and bracket style seems to be the most straight forward way to manage resources for streams.

I first implemented a version (by extending the Haskell RocksDB bindings) that managed iterators via GC by using `ForeignPtr`, but I wasn't able to implement database shutdown without major changes to the RocksDB Haskell API. 

(The python bindings seem to do GC for iterators, but I couldn't figure out the details of how they solve the races at db closure. I noticed the usage of weak references in several places in the python bindings, and decided that it wasn't worth it.)